sdks/js/DittoChatCore/src/slices/useMessages.ts

```diff
@@ -34,133 +38,356 @@ export interface MessageSlice {
-    }) => void,
+    onComplete: (result: AttachmentResult) => void,
   ) => void;
+  notificationHandler: ((message: MessageWithUser, room: Room) => void) | null;
+  registerNotificationHandler: (
+    handler: (message: MessageWithUser, room: Room) => void,
+  ) => void;
+  updateMessageReactions: (
+    message: Message,
+    room: Room,
+    reactions: Reaction[],
+  ) => Promise<void>;
+  addReactionToMessage: (
+    message: Message,
+    room: Room,
+    reaction: Reaction,
+  ) => Promise<void>;
+  removeReactionFromMessage: (
+    message: Message,
+    room: Room,
+    reaction: Reaction,
+  ) => Promise<void>;
 }
 
-interface ChatRetentionPolicy {
-  days: number;
+interface AttachmentResult {
+  success: boolean;
+  data?: Uint8Array;
+  metadata?: Record<string, string>;
+  error?: Error;
 }
 
+const DEFAULT_RETENTION_DAYS = 30;
+const MESSAGE_RECENCY_THRESHOLD = 10000; // 10 seconds
+const THUMBNAIL_MAX_SIZE = 282;
+
 export const createMessageSlice: CreateSlice<MessageSlice> = (
   _set,
   _get,
-  { ditto, userId, userCollectionKey }: DittoConfParams,
+  { ditto, userId, userCollectionKey, retentionDays: globalRetentionDays }
 ) => {
-  const chatRetentionPolicy: ChatRetentionPolicy = { days: 30 };
+  // Helper: Get room details
+  const getRoomDetails = async (room: Room) => {
+    if (!ditto) throw new Error("Ditto not initialized");
+
+    const result = await ditto.store.execute<Room>(
+      `SELECT * FROM ${room.collectionId || "rooms"} WHERE _id = :id`,
+      { id: room._id },
+    );
+
+    if (result.items.length === 0) throw new Error("Room not found");
+    return result.items[0].value;
+  };
+
+  // Helper: Get current user details
+  const getCurrentUser = async () => {
+    if (!ditto || !userId)
+      throw new Error("Ditto not initialized or user not found");
+
+    const result = await ditto.store.execute<ChatUser>(
+      `SELECT * FROM ${userCollectionKey} WHERE _id = :id`,
+      { id: userId },
+    );
+
+    const userValue = result.items?.[0]?.value;
+    return {
+      id: userId,
+      name: userValue?.name ?? userId,
+    };
+  };
+
+  const updateMessageLoadingState = () => {
+    _set((state: ChatStore) => {
+      return produce(state, (draft) => {
+        draft.messagesLoading = !draft.rooms.every(
+          (room) => room._id in draft.messagesByRoom,
+        );
+      });
+    });
+  };
+
+  // Helper: Handle message updates in state
+  const handleMessageUpdate = (
+    stateMessages: WritableDraft<MessageWithUser>[],
+    room: Room,
+    message: Message,
+    user?: ChatUser,
+  ) => {
+    const messageWithUser: MessageWithUser = {
+      message,
+      user,
+      id: message._id,
+    };
+
+    const mutableMessage = {
+      ...messageWithUser,
+      message: messageWithUser.message as unknown as WritableDraft<
+        typeof messageWithUser.message
+      >,
+      user: messageWithUser.user as unknown as WritableDraft<
+        typeof messageWithUser.user
+      >,
+    };
```

Comment:
These casts should be handled by immer IMO, they have `castDraft` for just that:

```suggestion
    const mutableMessage = castDraft({
      ...messageWithUser,
      message: messageWithUser.message,
      user: messageWithUser.user,
    });
```

I understand that the implementation of `castDraft` basically does the same, though. This felt cleaner

Reply:
Changed to castDraft